frame/vamm/src/tests/swap_asset.rs

```diff
@@ -0,0 +1,477 @@
+					invariant,
+					closed: None,
+				},
+			),
+		],
+	}
+	.build()
+	.execute_with(|| {
+		let swap = TestPallet::swap(&swap_config);
+		let vamm_after_swap = VammMap::<MockRuntime>::get(0);
+		let vamm_expected = VammMap::<MockRuntime>::get(1);
+
+		assert_eq!(swap, Ok(39_215_686_275));
+		assert_eq!(vamm_after_swap, vamm_expected);
+	})
+}
+
+// TODO(Cardosaum): fix code when try to remove quote.
+// as of now all calls for this instance return Ok(0), which essentially would
+// be a no-op.
+#[test]
+#[ignore = "to be implemented"]
+fn swap_remove_quote() {
+	let base_u256 = U256::from(2_u128) * U256::exp10(12);
+	let quote_u256 = U256::from(50_u128) * U256::exp10(12);
+	let invariant = base_u256 * quote_u256;
+
+	let swap_config = SwapConfig {
+		vamm_id: 0,
+		asset: AssetType::Quote,
+		input_amount: base_u256.as_u128() / 100_u128,
+		direction: Direction::Remove,
+		output_amount_limit: 0,
+	};
+
+	ExtBuilder {
+		vamm_count: 1,
+		vamms: vec![
+			// Initial state
+			(
+				0,
+				VammState {
+					base_asset_reserves: base_u256.as_u128(),
+					quote_asset_reserves: quote_u256.as_u128(),
+					peg_multiplier: 1,
+					invariant,
+					closed: None,
+				},
+			),
+			// Expected final state
+			(
+				1,
+				VammState {
+					base_asset_reserves: 1_960_784_313_725,
+					quote_asset_reserves: 51_000_000_000_000,
+					peg_multiplier: 1,
+					invariant,
+					closed: None,
+				},
+			),
+		],
+	}
+	.build()
+	.execute_with(|| {
+		// let vamm_before_swap = VammMap::<MockRuntime>::get(0);
+		let swap = TestPallet::swap(&swap_config);
+		let vamm_after_swap = VammMap::<MockRuntime>::get(0);
+		let vamm_expected = VammMap::<MockRuntime>::get(1);
+
+		assert_eq!(swap, Ok(39_215_686_275));
+		assert_eq!(vamm_after_swap, vamm_expected);
+	})
+}
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(1))]
+	#[test]
+	fn multiple_swaps_dont_diverge_from_original_invariant_both_base_and_quote(
+		mut vamm_state in any_vamm_state(),
+		swap_config in multiple_swaps()
+	) {
+		// Ensure vamm is always open
+		vamm_state.closed = None;
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			let vamm_before_swap = VammMap::<MockRuntime>::get(0);
+			for x in swap_config.iter() {
+				assert_ok!(TestPallet::swap(x));
+			}
+			let vamm_after_swap = VammMap::<MockRuntime>::get(0);
+
+			let invariant_before = TestPallet::compute_invariant(
+				vamm_before_swap.unwrap().base_asset_reserves,
+				vamm_before_swap.unwrap().quote_asset_reserves,
+			).unwrap();
+
+			let invariant_after = TestPallet::compute_invariant(
+				vamm_after_swap.unwrap().base_asset_reserves,
+				vamm_after_swap.unwrap().quote_asset_reserves,
+			).unwrap();
+
+			let invariant_epsilon = invariant_before / U256::exp10(8);
+			let invariant_delta = invariant_before.max(invariant_after)
+				- invariant_before.min(invariant_after);
+
+			assert_eq!(invariant_delta / vamm_state.invariant, U256::zero());
```

Comment:
does this default/round to zero if invariant_delta < vamm_state.invariant? trying something like `assert_eq!(1u128 / 2u128, 0u128)` is true even though it shouldn't necessarily be

Reply:
Good catch!

Actually I forgot to remove those `assert_eq!` as they are doing essentially nothing since we have `assert!(invariant_delta <= invariant_epsilon)` in the next line (for the 3 instances of this test: both assets, only base and only quote).

I removed those unnecessary asserts, thanks for the input! :)